jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentSourceRange.java

```diff
@@ -0,0 +1,194 @@
+     * as a sub-range of the original.
+     *
+     * @param source The {@link Content.Source} to wrap.
+     * @param offset the offset byte of the content to start from.
+     *               Must be greater than or equal to 0 and less than the content length (if known).
+     * @param length the length of the content to make available, -1 for the full length.
+     *               If the size of the content is known, the length may be truncated to the content size minus the offset.
+     * @throws IndexOutOfBoundsException if the offset or length are out of range.
+     * @see TypeUtil#checkOffsetLengthSize(long, long, long)
+     */
+    public ContentSourceRange(Content.Source source, long offset, long length)
+    {
+        this(source, offset, length, false);
+    }
+
+    /**
+     * Create a {@link ContentSourceRange} which wraps another {@link Content.Source} to appear
+     * as a sub-range of the original.
+     *
+     * @param source The {@link Content.Source} to wrap.
+     * @param offset the offset byte of the content to start from.
+     *               Must be greater than or equal to 0 and less than the content length (if known).
+     * @param length the length of the content to make available, -1 for the full length.
+     *               If the size of the content is known, the length may be truncated to the content size minus the offset.
+     * @param readToEof - if true, read until EOF of the source after the range is read.
+     * @throws IndexOutOfBoundsException if the offset or length are out of range.
+     * @see TypeUtil#checkOffsetLengthSize(long, long, long)
+     */
+    public ContentSourceRange(Content.Source source, long offset, long length, boolean readToEof)
+    {
+        _source = source;
+        _offset = offset;
+        _length = TypeUtil.checkOffsetLengthSize(offset, length, source.getLength());
+        _offsetRemaining = _offset;
+        _lengthRemaining = _length;
+        _readToEof = readToEof;
+    }
+
+    @Override
+    public long getLength()
+    {
+        return _length;
+    }
+
+    @Override
+    public Content.Chunk read()
+    {
+        if (_terminal != null)
+        {
+            Content.Chunk chunk = _terminal;
+            _terminal = Content.Chunk.next(_terminal);
+            return chunk;
+        }
+
+        while (true)
+        {
+            Content.Chunk chunk = _source.read();
+            if (chunk == null)
+                return null;
+
+            if (Content.Chunk.isFailure(chunk))
+            {
+                _terminal = Content.Chunk.next(chunk);
+                return chunk;
+            }
+
+            if (_offsetRemaining > 0)
+            {
+                if (_offsetRemaining >= chunk.remaining())
+                {
+                    // We can skip this whole chunk.
+                    _offsetRemaining -= chunk.remaining();
+                    chunk.clear();
```

Comment:
No need to clear?

Reply:
I agree there is no need to clear, but not for the reason you give.